go-controller/pkg/mock/mock_lsp.go

```diff
@@ -0,0 +1,301 @@
+	return nil, fmt.Errorf("method %s is not implemented yet", functionName())
+}
+
+// Get Options for LSP
+func (mock *MockOVNClient) LSPGetOptions(lsp string) (map[string]string, error) {
+	return nil, fmt.Errorf("method %s is not implemented yet", functionName())
+}
+
+// Set dynamic addresses in LSP
+func (mock *MockOVNClient) LSPSetDynamicAddresses(lsp string, address string) (util.OVNCommandInterface, error) {
+	cachedErr := mock.retFromErrorCache(LogicalSwitchPortType, lsp, LogicalSwitchPortDynamicAddresses)
+	if cachedErr != nil {
+		return nil, cachedErr
+	}
+
+	var updatelspCache UpdateObjectCacheByName
+	var ok bool
+	if updatelspCache, ok = mock.updateTransactionCache[LogicalSwitchPortType]; !ok {
+		updatelspCache = make(UpdateObjectCacheByName)
+		updatelspCache[lsp] = make([]UpdateCache, 0)
+	}
+	updates := updatelspCache[lsp]
+
+	updates = append(updates, UpdateCache{
+		FieldType:  LogicalSwitchPortDynamicAddresses,
+		FieldValue: address,
+	})
+	updatelspCache[lsp] = updates
+
+	mock.updateTransactionCache[LogicalSwitchPortType] = updatelspCache
+
+	return &MockOVNCommand{Exe: mock}, nil
+}
+
+// Get dynamic addresses from LSP
+func (mock *MockOVNClient) LSPGetDynamicAddresses(lsp string) (string, error) {
+	lspRet, err := mock.LSPGet(lsp)
+	if err != nil {
+		return "", err
+	}
+	if lspRet != nil {
+		return "", fmt.Errorf("no lsp found with name: %s", lsp)
+	}
+	return lspRet.DynamicAddresses, nil
+}
+
+// Set external_ids for LSP
+func (mock *MockOVNClient) LSPSetExternalIds(lsp string, external_ids map[string]string) (util.OVNCommandInterface, error) {
+	cachedErr := mock.retFromErrorCache(LogicalSwitchPortType, lsp, LogicalSwitchPortExternalId)
+	if cachedErr != nil {
+		return nil, cachedErr
+	}
+	var updatelspCache UpdateObjectCacheByName
+	var ok bool
+	if updatelspCache, ok = mock.updateTransactionCache[LogicalSwitchPortType]; !ok {
+		updatelspCache = make(UpdateObjectCacheByName)
+		updatelspCache[lsp] = make([]UpdateCache, 0)
+	}
+	updates := updatelspCache[lsp]
+	updates = append(updates, UpdateCache{
+		FieldType:  LogicalSwitchPortExternalId,
+		FieldValue: external_ids,
+	})
+
+	updatelspCache[lsp] = updates
+	mock.updateTransactionCache[LogicalSwitchPortType] = updatelspCache
+
+	return &MockOVNCommand{Exe: mock}, nil
+}
+
+// Get external_ids from LSP
+func (mock *MockOVNClient) LSPGetExternalIds(lsp string) (map[string]string, error) {
+	lspRet, err := mock.LSPGet(lsp)
+	if err != nil {
+		return nil, err
+	}
+	if lspRet != nil {
+		return nil, fmt.Errorf("no lsp found with name: %s", lsp)
+	}
+	extIds := make(map[string]string)
+	for k, v := range lspRet.ExternalID {
+		key, keyOk := k.(string)
+		value, valueOk := v.(string)
+		if !keyOk || !valueOk {
+			continue
+		}
+		extIds[key] = value
+	}
+	return extIds, nil
+}
+
+func (mock *MockOVNClient) updateLSPCache(updates UpdateObjectCacheByName, mockCache MockObjectCacheByName) error {
```

Comment:
I don't think this is right; as it's not correctly reentrant. Specifically, if I do

```
cmd1 := AddLSP()
cmd2 := AddLSP()
Execute(cmd2)
```

I'll still see the effects of cmd1.

I think you need to move this to a closure.

Reply:
Yeah, you are right. The `Execute()` totally ignores the cmd argument and applies everything in the cache at that point in time. We sure can get fancier with the mock and re-write it to keep track of the individual commands' operations with a closure, but is it really worth the effort for a mock?

Reply:
It's less code and more correct, so it seems worth it to me.